### ISSUE TITLE
Collection expressions: Allow builder-pattern method to return different type than the type being built.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/csharplang#7396]
This issue contains a change request to the current [collection expressions](https://github.com/dotnet/csharplang/blob/main/proposals/collection-expressions.md) specification that the working group would like to bring for LDM consideration.

The title, but the core use case we want to support is relates to the following APIs in the BCL.  First, these types, with these members exist today:

```c#
public interface IImutableSet<T> { /* ... */ }

[CollectionBuilder(typeof(ImmutableHashSet), "Create")]
public class ImmutableHashSet<T> : IImutableSet<T> { /* ... */  }

public static class ImmutableHashSet
{
    public static ImmutableHashSet<T> Create<T>(ReadOnlySpan<T> values);
}
```

As such, given these canonical implementations of these interfaces, we would like to allow a user to write code like so:

```c#
public class C
{
    public IImutableSet<X> Values { get; } = [GetValue(a), GetValue(b), GetValue(c)];
}
```

This doesn't work because `IImutableSet<T>` is currently not something the compiler knows how to create.  This is unfortunate given such a canonical implementation.  Fortunately, a solution seems readily available.  Specifically, update the interface like so:

```diff
+ [CollectionBuilder(typeof(ImmutableHashSet), "Create")]
public interface IImutableSet<T> { /* ... */ }
```

Bringing it in line with the class.  With this, the compiler would now be able to find `ImmutableHashSet.Create` as the way to initialize `C.Values`.   *However*, as the spec is currently written, this would not work. Specifically, the [specification says](https://github.com/dotnet/csharplang/blob/main/proposals/collection-expressions.md#create-methods): 

> For the create method:
> - ...
> - There is an [identity conversion](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/conversions.md#1022-identity-conversion) between the method return type and the collection type.

Here the method return type is `ImmutableSet<T>` (the concrete type) while the collection type is `IImutableSet<T>` (the interface type).   This is not an identity conversion, so this is not allowed.

However, given how trivial this is to support, and given that the BCL is fine adding these attributes to these collections ([see PR here](https://github.com/dotnet/runtime/pull/89459)), we propose tweaking the above rule to one of either:

1. There is an *implicit conversion* between the method return type and the collection type.

In other words, allow the user's code above because it would be equivalent to:

```c#
public class C
{
    public IImutableSet<X> Values { get; } = ImmutableHashSet.Create(GetValue(a), GetValue(b), GetValue(c));
}
```

Given that the full form would be legal to write there, requiring an exact implicit conversion, instead of any implicit conversion is just onerous.

Costs here are minimal.  BCL PR is already done, and compiler PR would just be a minor tweak to the check, along with tests to validate behavior.
